apps/studio/src/lib/db/clients/BasicDatabaseClient.ts

```diff
@@ -30,42 +34,78 @@ export interface AppContextProvider {
 }
 
 // raw result type is specific to each database implementation
-export abstract class BasicDatabaseClient<RawResultType> implements DatabaseClient {
+export abstract class BasicDatabaseClient<RawResultType> {
 
   knex: Knex | null;
   contextProvider: AppContextProvider
+  server: IDbConnectionServer;
+  database: IDbConnectionDatabase;
+  connectionType: ConnectionType;
 
-
-  constructor(knex: Knex | null, contextProvider: AppContextProvider) {
+  constructor(knex: Knex | null, contextProvider: AppContextProvider, server: IDbConnectionServer, database: IDbConnectionDatabase) {
     this.knex = knex;
     this.contextProvider = contextProvider
-  }
-  listTablePartitions(_table: string, _schema: string): Promise<TablePartition[]> {
-    return Promise.resolve([])
-  }
-  alterPartitionSql(_changes: AlterPartitionsSpec): string {
-    return ''
+    this.server = server;
+    this.database = database;
+    this.connectionType = this.server?.config.client;
   }
 
-  async alterPartition(_changes: AlterPartitionsSpec): Promise<void> {
-    return;
-  }
+  abstract getBuilder(table: string, schema?: string): ChangeBuilderBase
 
-  async getMaterializedViewCreateScript(_view: string, _schema?: string): Promise<string[]> {
-    return [];
-  }
-  
+  // DB Metadata ****************************************************************
+  abstract supportedFeatures(): SupportedFeatures;
   abstract versionString(): string;
-
   defaultSchema(): string | null {
     return null
   }
+  // ****************************************************************************
 
-  abstract getBuilder(table: string, schema?: string): ChangeBuilderBase
+  // Connection *****************************************************************
+  async connect(): Promise<void> {
+    /* eslint no-param-reassign: 0 */
+    if (this.database.connecting) {
+      throw new Error('There is already a connection in progress for this database. Aborting this new request.');
+    }
 
-  abstract supportedFeatures(): SupportedFeatures;
-  abstract connect(): Promise<void>
-  abstract disconnect(): Promise<void>;
+    try {
+      this.database.connecting = true;
+
+      // terminate any previous lost connection for this DB
+      if (this.database.connected) {
+        this.disconnect();
+      }
+
+      // reuse existing tunnel
+      if (this.server.config.ssh && !this.server.sshTunnel) {
```

Comment:
So here we only use `config` from `server`. Calling it `server` has always been confusing to me, shouldn't we just accept the `config` portion of it?